### PR TITLE
Add mechanism to deduplicate multiple triggers

### LINKED ISSUE #48: Support conversion deduplication in Event-Level API
Currently, the aggregate API explainer proposes a new attribute, [`local-dedup-key`](https://github.com/WICG/conversion-measurement-api/blob/master/AGGREGATE.md#many-per-click--once-per-click), for conversion registration that allows the reporting origin to deduplicate conversions client side.

In the Event-level API it is already possible for the reporting origin to do server side de-duping by looking at conversion-metadata, assuming that those bits are used in a way that de-duping is possible.

It may be possible to support this attribute in the event-level API, as it should not introduce any new information leakage on the conversion side. This would allow for a unified way of de-duping conversions client side, as well as allow the conversion metadata to be used for things other than de-duping.

### PR DESCRIPTION
**Pull request by johnivdel**
Fixes #48

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -190,9 +190,11 @@ this API:
 to trigger attribution for all matching sources.
 
 The browser will treat redirects to a URL of the form:
-`https://<attributionreportto>/.well-known/attribution-reporting/trigger-attribution[?data=<data>]`
+`https://<attributionreportto>/.well-known/attribution-reporting/trigger-attribution[?data=<data>&dedup-key=<dedup-key>]`
 
-as a special request, where optional data associated with the event that triggered attribution is stored in a query parameter.
+as a special request. The query string for the request contains additional information about the attributeion trigger:
```
Comment (apasel422) [inline README.md:195]:
```suggestion
as a special request. The query string for the request contains additional information about the attribution trigger:
```

File: README.md [hunk 2]
```diff
@@ -231,9 +233,11 @@ The most recent matching source is given a `credit` of value 100. All other matc
 
 For each matching source, schedule a report. To schedule a report,
 the browser will store
- {`attributionreportto`, `attributiondestination` eTLD+1, `attributionsourceeventid`, [decoded](#data-encoding) trigger-data, credit} for the source.
+ {`attributionreportto`, `attributiondestination` eTLD+1, `attributionsourceeventid`, [decoded](#data-encoding) trigger-data, credit, dedup-key} for the source.
 Scheduled reports will be sent as detailed in [Sending scheduled reports](#sending-scheduled-reports).
 
+The browser will only create reports for a source, the trigger's dedup-key has not already been associated with a report for that source.
```
Comment (apasel422) [inline README.md:239]:
```suggestion
The browser will only create reports for a source if the trigger's dedup-key has not already been associated with a report for that source.
```

**Comment by csharrison** [comment]
Mention that this fixes https://github.com/WICG/conversion-measurement-api/issues/48

**Review by csharrison — APPROVED** [review]

**Comment by apasel422** [comment]
Is the dedup key an integer, an arbtirary string, something else?

File: README.md [hunk 3]
```diff
@@ -231,9 +233,11 @@ The most recent matching source is given a `credit` of value 100. All other matc
 
 For each matching source, schedule a report. To schedule a report,
 the browser will store
- {`attributionreportto`, `attributiondestination` eTLD+1, `attributionsourceeventid`, [decoded](#data-encoding) trigger-data, credit} for the source.
+ {`attributionreportto`, `attributiondestination` eTLD+1, `attributionsourceeventid`, [decoded](#data-encoding) trigger-data, credit, dedup-key} for the source.
 Scheduled reports will be sent as detailed in [Sending scheduled reports](#sending-scheduled-reports).
 
+The browser will only create reports for a source if the trigger's dedup-key has not already been associated with a report for that source.
```
Comment (apasel422) [inline README.md:239]: Does this imply that the browser stores every dedup-key value for every source forever? Or is the dedup-key only valid for a browser-determined period?

Comment (johnivdel) [reply README.md:239]: The dedup-keys can only deduplicate triggers for the source they are attributed to. The browser only needs to keep it around for the lifetime of the source.

So for example, if we have:

source A
trigger 1, dedup=1
trigger 2, dedup=1
source B
trigger 3, dedup=1

The browser will send 2 reports:

<source A, trigger 1>
<source B, trigger 3>

**Comment by johnivdel** [comment]
> Is the dedup key an integer, an arbitrary string, something else?

Clarified that the dedup-key is an integer. We could use strings, but this way we have a simple upperbound on length.

**Review by apasel422 — APPROVED** [review]

**Comment by apasel422** [comment]
Is there anything left to address here, or can we merge?

File: event_attribution_reporting_clicks.md [hunk 4]
```diff
@@ -191,12 +188,12 @@ this API:
...
+`https://<attributionreportto>/.well-known/attribution-reporting/trigger-attribution[?data=<data>&priority=<priority>&dedup-key=<dedup-key>]`
 
-as a special request, where optional data associated with the event that triggered attribution is stored in a query parameter.
-
-- `data`: an integer label used for the trigger, limited to 3 bits.
-- `priority`: a signed 64-bit integer representing the priority of this trigger compared to other triggers for the same source.
+as a special request. The query string for the request contains additional information about the attribution trigger:
+- `data`: optional data to identify the triggering event
+- `priority`: optional signed 64-bit integer representing the priority of this trigger compared to other triggers for the same source.
+- `dedup-key`: optional signed 64 bit integer which will be used to deduplicate multiple triggers which contain the same dedup-key for a single source
```
Comment (pwnall) [inline event_attribution_reporting_clicks.md:196]: nit: `64 bit` is spelled inconsistently (line 195 has it as a `64-bit`)

Comment (johnivdel) [reply event_attribution_reporting_clicks.md]: Fixed, thanks!

**Pull request merged**